tests/unit/test_unique_any.cpp

```diff
@@ -0,0 +1,394 @@
+    // be in a moved from state, and enforcing std::move(a) it
+    // is made clearer in the calling code that a has been invalidated.
+    //     util::any_cast<T&&>(a)
+    T val(util::any_cast<T&&>(std::move(a)));
+    EXPECT_EQ(val.value, 100);
+}
+
+// test any_cast(unique_any*) and any_cast(const unique_any*)
+//   - these have different behavior to any_cast on reference types
+//   - are used by the any_cast on refernce types
+TEST(unique_any, any_cast_ptr) {
+    using util::unique_any;
+
+    // test that valid pointers are returned for int and std::string types
+
+    unique_any ai(42);
+    auto ptr_i = util::any_cast<int>(&ai);
+    EXPECT_EQ(*ptr_i, 42);
+
+    unique_any as(std::string("hello"));
+    auto ptr_s = util::any_cast<std::string>(&as);
+    EXPECT_EQ(*ptr_s, "hello");
+
+    // test that exceptions are thrown for invalid casts
+    EXPECT_EQ(util::any_cast<int>(&as), nullptr);
+    EXPECT_EQ(util::any_cast<std::string>(&ai), nullptr);
+    unique_any empty;
+    EXPECT_EQ(util::any_cast<int>(&empty), nullptr);
+    EXPECT_EQ(util::any_cast<int>((util::unique_any*)nullptr), nullptr);
+
+    // Check that constness of the returned pointer matches that the input.
+    {
+        unique_any a(42);
+        auto p = util::any_cast<int>(&a);
+        static_assert(std::is_same<int*, decltype(p)>::value,
+                "any_cast(any*) should not return const*");
+    }
+    {
+        const unique_any a(42);
+        auto p = util::any_cast<int>(&a);
+        static_assert(std::is_same<const int*, decltype(p)>::value,
+                "any_cast(const any*) should return const*");
+    }
+}
+
+// test any_cast(unique_any&)
+TEST(unique_any, any_cast_ref) {
+    util::unique_any ai(42);
+    auto& i = util::any_cast<int&>(ai);
+
+    EXPECT_EQ(typeid(i), typeid(int));
+    EXPECT_EQ(i, 42);
+
+    // any_cast<T>(unique_any&) returns a 
+    i = 100;
+    EXPECT_EQ(util::any_cast<int>(ai), 100);
+}
+
+// test any_cast(const unique_any&)
+TEST(unique_any, any_cast_const_ref) {
+    const util::unique_any ai(42);
+    auto& i = util::any_cast<const int&>(ai);
+
+    EXPECT_EQ(typeid(i), typeid(int));
+    EXPECT_EQ(i, 42);
+
+    EXPECT_TRUE((std::is_same<const int&, decltype(i)>::value));
+}
+
+// test any_cast(any&&)
+TEST(unique_any, any_cast_rvalue) {
+    auto moved = util::any_cast<moveable>(util::unique_any(moveable()));
+    EXPECT_EQ(moved.moves, 2);
+    EXPECT_EQ(moved.copies, 0);
+}
+
+TEST(unique_any, std_swap) {
+    util::unique_any a1(42);
+    util::unique_any a2(3.14);
+
+    auto pi = util::any_cast<int>(&a1);
+    auto pd = util::any_cast<double>(&a2);
+
+    std::swap(a1, a2);
+
+    // test that values were swapped
+    EXPECT_EQ(util::any_cast<int>(a2), 42);
+    EXPECT_EQ(util::any_cast<double>(a1), 3.14);
+
+    // test that underlying pointers did not change
+    EXPECT_EQ(pi, util::any_cast<int>(&a2));
+    EXPECT_EQ(pd, util::any_cast<double>(&a1));
+}
+
+// test operator=(any&&)
```

Comment:
Test comment out of date: this is testing `unique_any::operator=(unique_any&&)`.